src/sources/audiosource.cpp

```diff
@@ -27,6 +27,7 @@ AudioSource::AudioSource(
           m_signalInfo(signalInfo),
           m_bitrate(inner.m_bitrate),
           m_frameIndexRange(inner.m_frameIndexRange) {
+    DEBUG_ASSERT(m_frameIndexRange.length() >= 0);
```

Comment:
What do you mean with "technically UB"? 

This basically asserts the trivial condition here: 
https://github.com/mixxxdj/mixxx/blob/4a51c393eb8820a5728f4b2b53dce8ef203d2b8c/src/util/indexrange.h#L68
My guess is that the I originally mean to assert that the range is not backward like in `AudioSource::initFrameIndexRangeOnce()`
So I think it is reasonable to replace it with a check that range is not backward.

Reply:
> What do you mean with "technically UB"?

No, I mean integer over-/underflow is only defined behavior on unsigned integral types. So there is no reliable way of detecting them just by inspecting the value (strictly theoretically speaking). The only way would be to check after every arithmetic operation (feasible by using an abstraction such as [SafeInt](https://github.com/dcleblanc/SafeInt)). 

But that's just my theoretical two cents on the matter and not a change I propose for this PR. 

> My guess is that the I originally mean to assert that the range is not backward like in AudioSource::initFrameIndexRangeOnce()

I apologize, the location of the comment was chosen suboptimally. The concern rather applies to the changes in the `Analyzer*` classes such as here:

https://github.com/mixxxdj/mixxx/pull/11511/commits/ae94ca7d8a4fcfecfa2fb1f6fbaeb9f4f1fd1bcc#diff-d76237758c9ef4a039bb906e468962e074a9464f40a0219a273d4263523ed88fR47

Reply:
The link cites `AnalyzerBeats::initialize()`
```
    if (frameLength <= 0) {
        return false;
    }
```
There is noting wrong with checking the the input `frameLength` here. 

> So there is no reliable way of detecting them just by inspecting the value (strictly theoretically speaking). 

This is the case for signed AND unsigned values, because the overflowed value can be a valid value in an expected range in both cases. So we are here on the same page IMHO.  

The undefined signed int overflow is fortunately only true for C++ itself. The int overflow behavior is well defined by our current target architectures. I know there are some exotic target architectures out there, where this is not the case, but they will never be a Mixxx build target. I am also in doubt that Qt will run on them. 
Can we put a statement in our style guide or implement a unit test for in overflow. Than we can cite it next time.

Reply:
> There is noting wrong with checking the the input frameLength here.

Yes, in this particular case its fine since we're checking the length of something small, but we absolutely can't be doing that for detecting overflow.

> This is the case for signed AND unsigned values, because the overflowed value can be a valid value in an expected range in both cases. So we are here on the same page IMHO.

Well since this is a count of samples, we expect negative values to be unexpected right? and that's why we're checking for `<= 0`. 

> The undefined signed int overflow is fortunately only true for C++ itself. The int overflow behavior is well defined by our current target architectures. I know there are some exotic target architectures out there, where this is not the case, but they will never be a Mixxx build target. I am also in doubt that Qt will run on them.

That does not matter. Its UB in C++ and thus compilers will make use of that for optimization. Here are two examples where the compiler optimizes based on the non-existence of overflow: https://godbolt.org/z/zezfcxWxE

Reply:
I realize that I'm complaining about an issue that does not apply to this PR, so we can stop here now if you prefer to. Just to summarize, checking whether integer overflow occurred by inspecting a value (as it looks like on first sight in https://github.com/mixxxdj/mixxx/pull/11511/commits/a5ddb48a572baf5fc353425fdcc24fc6ae951d89 judging by the commit message) will not work. Can we agree on that?

Reply:
> Can we put a statement in our style guide or implement a unit test for in overflow. 

Whats the statement supposed to be? "Don't rely on undefined behavior"?

Reply:
> Just to summarize, checking whether integer overflow occurred by inspecting a value (as it looks like on first sight in https://github.com/mixxxdj/mixxx/commit/a5ddb48a572baf5fc353425fdcc24fc6ae951d89 judging by the commit message) will not work. Can we agree on that?

Yes.

> Whats the statement supposed to be? "Don't rely on undefined behavior"?

Something like that: 

Note: Even though the encoding of signed integer is not defined by the C++ standard, the two's complement representation of signed numbers is enforced by the Qt library: https://github.com/qt/qtbase/blame/6.5/src/corelib/global/qtypes.cpp#L362  
This means you can rely that for instance a a int8_t of 127 will become -128 after increment by one. There is no need to consider other singed encodings.

Reply:
> Note: Even though the encoding of signed integer is not defined by the C++ standard, the two's complement representation of signed numbers is enforced by the Qt library: https://github.com/qt/qtbase/blame/6.5/src/corelib/global/qtypes.cpp#L362

The linked static assert does not check whether integer overflow is valid. It just checks the boundaries of `int` and infers that they're encoded in two's complement. It says nothing about overflowing behavior!!

Reply:
Ah right. Here is the original commit https://github.com/qt/qtbase/commit/34ae919f23b7d89844358afea23c73dc6469faf3
The overflow behavior is still undefined but the possible options are narrowed. Since we are at C++20 now and the undefined option is the fastest. There is noting to do.